### PR TITLE
Discard errors from browser extensions if ignore3rdPartyErrors option is true

### PR DESCRIPTION
**Pull request by samuel-holt**
## Overview

If an error occurs in a script that is injected into the body of the website's document by an extension, this error can be caught by the `window.onerror` handler. 

From researching this, I've found errors from browser extensions can take on multiple forms. This PR covers three particular cases, with the aim that if a developer wants to ignore third-party errors, these errors should also be ignored.

If the error occurs within a function, then each line of the stacktrace will have a `url` value prefixed with `*-extension://` (where `*` is one of 'chrome', 'safari', 'safari-web', or 'moz').

For example:

```js
{
  message: 'Error in a browser extension function',
  stack: [
    {
      line: 12,
      column: 345,
      url: 'chrome-extension://example/index.js',
      func: 'render'
    },
  ]
}
```

There is now a utility function for detecting browser extension errors like this. This function could be used for specifically ignoring errors from browser extensions in future.

An error can occur directly in an injected script, this can lead to a stacktrace with null values like so:

```js
{
  message: 'Error in a browser extension script',
  stack: [
    {
      line: null,
      column: null,
      url: null,
      func: 'onLoadHandler'
    },
  ]
}
```

I've observed another more obscure case, where the `url` is set to the current website's URL, and the `func` is '?'.

```js
{
  message: 'Error from a browser extension, bot or crawler',
  stack: [
    {
      line: 0,
      column: 0,
      url: 'http://example.com/index.html',
      func: '?'
    },
  ]
}
```

There's a utility function to detect if a stacktrace matches this pattern, and the error will be discarded if every line in the stacktrace matches.

## Updates

- Add a new error utility script with functions for detecting various types of error
- Move some of the existing logic to the utility script for easier unit testing
- Extact the `ignore3rdPartyErrors` logic out into a separate function 
- Add helper utilities for various common functions

### CODE REVIEW

File: src/raygun.utilities/errorUtilities.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
+/**
+ * @prettier
+ */
+
+window.raygunErrorUtilitiesFactory = function (window, Raygun) {
```
Comment (krishnaKapadia) [inline src/raygun.utilities/errorUtilities.js:5]: Would be good to handle the case where what about the case where this is run in an non-browser environment?

**Review by krishnaKapadia — COMMENTED** [review]
Another question that we sort of touched on in the Mob review

**Review by izsi-salmon — DISMISSED** [review]
🌮

File: src/raygun.utilities/errorUtilities.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+      return utils.any(stack, function (stackLine) {
+        var url = stackLine.url;
+
+        return !utils.isNil(url) && isBrowserExtensionUrl(url);
+      });
+    },
+
+    /**
+     * Check if all lines in the stack are invalid, i.e. they have a line and column number of zero and a url of null or
```
Comment (JasenPalmer) [inline src/raygun.utilities/errorUtilities.js:88]: Perhaps it would be more clear to say "Check if any of the lines in the stack are valid"

**Review by JasenPalmer — APPROVED** [review]
👍🏻

**Review by krishnaKapadia — APPROVED** [review]
🚀 Nice work on this one!

**Pull request merged**